### PR TITLE
reduce cost of large variant matrix

### PR DESCRIPTION
**Pull request by minrk**
### Description

when variant matrix is large and mostly unused (as in conda-forge), the length of input_variants may be several thousand (13,824 in the case of petsc4py) when only a few are actually used.

This causes `get_loop_vars` and `metadata.copy()` to become very expensive and dominate render time.

This reduction cuts time spent in `render_recipe` for petsc4py from over 2 minutes to 40 seconds to produce 72 actual variants:

before:

Screenshot 2024-06-28 at 13 43 28

after:

Screenshot 2024-06-28 at 13 43 50

(result is unchanged)

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by minrk** [comment]
need to investigate why the conda-build tests produce different results with actual runs of conda-build, all of which seem to produce the right variants.

**Comment by minrk** [comment]
seems to be the exclusion of `get_used_vars` doesn't actually cover all used variables. And some tests appear to expect matrix entries for explicitly unused variables (bzip2), which makes me think that perhaps this should be a conda-smithy thing and not a conda-build thing.

**Comment by beeankha** [comment]
pre-commit.ci autofix

**Comment by minrk** [comment]
Marking as draft because I'm not so sure about the implications of reducing the variant matrix, and this is perhaps too fragile and complicated to work on.

**Comment by minrk** [comment]
I've taken a different approach that doesn't modify the variants list at all, so shouldn't have any consequences besides performance. Instead of reducing the actual variants list, I've reduced the cost of the two dominant operations on the large variant list:

- defer copying `metadata.config.variants` for each top_level variant until after it's been filtered, so far fewer variant dicts are copied (each variant should be copied exactly once, I believe, rather than once per top_level loop)
- in `get_used_loop_vars`, compute `used_vars` first and pass it to `get_loop_vars` so only the used subset are considered, rather than comparing all keys across all variants every time.

The first changes the variants copy from O(top level variants * input_variants) to O(top_level_variants * per_top_level_variants), which is the same as `O(input_variants)`. So the reduction is equal to the number of top level variants; in the case of petsc4py on linux-64, that's `72 * 13,824 = 995,328`, reduced to `72 * 192 = 13,824`, a reduction of 98.6%. In the case of conda-forge, the _real_ used number of variants is 72, so there's still a _further_ 99.5% that the original proposal reduced by, but I don't know how to do that safely while maintaining all the unspecified guarantees in conda-build.

The second changes the `get_loop_vars` call from O(variant keys * input variants) to O(used vars * input variants), in the case of petsc4py that's `~365 -> 14`, a reduction of 96%:

The savings aren't _quite_ what they are for actually reducing the variants list because there are still some operations on the full list but it still cuts render time in half.

**Comment by minrk** [comment]
I'm afraid I don't understand the mac failures or how they could be related to this PR. Hopefully something transient?

**Comment by beckermr** [comment]
Those mac failures do look unrelated. Let's try to rerun before we dig into them.

**Comment by beckermr** [comment]
@mbargull Can you take a look at this one? I don't see any obvious problems but this work gets a bit into the guts of the code.

**Comment by minrk** [comment]
Anyone have a chance to look at this?

File: conda_build/render.py [hunk 1]
```diff
@@ -860,6 +869,11 @@ def distribute_variants(
                 )
                 or mv.config.variants
             )
+        # copy variants before we start modifying them,
+        # but after we've reduced the list via the conform_dict filter
+        mv.config.variants = pickle.loads(
+            pickle.dumps(mv.config.variants, pickle.HIGHEST_PROTOCOL)
+        )
```
Comment (isuruf) [inline conda_build/render.py:876]: Should we just add a `copy_variants` method to config to reduce doing the same thing as https://github.com/conda/conda-build/blob/adfdc2a43d968c934d14bbaa30b1a023f68a77fc/conda_build/config.py#L827-L829 ?

Comment (minrk) [reply conda_build/render.py]: good idea, added.

**Review by beckermr — APPROVED** [review]

**Comment by beckermr** [comment]
@isuruf @beeankha @kenodegard Can we merge this one for the 2.49 release?

File: news/5392-variant-copy [hunk 2]
```diff
@@ -0,0 +1,3 @@
+### Enhancements
+
+* reduce render time when there is a large number of unused variants
```
Comment (beeankha) [inline news/5392-variant-copy:3]:
```suggestion
* Reduce render time when there is a large number of unused variants. (#5392)
```

**Review by beeankha — DISMISSED** [review]

**Review by beeankha — APPROVED** [review]

**Review by beeankha — APPROVED** [review]

**Pull request merged**